src/lib.rs

```diff
@@ -1,103 +1,163 @@
+#[derive(Debug, Serialize, PartialEq, Eq)]
+pub struct SendgridEmail<'a> {
+    #[serde(skip)]
+    api_key: String,
+
+    #[serde(rename = "personalizations")]
+    personalizations: Vec<Personalization>,
+
+    #[serde(rename = "from")]
+    from: From,
+
+    #[serde(rename = "subject")]
+    subject: Option<String>,
+
+    #[serde(rename = "content")]
+    content: Vec<Content<'a>>,
+}
+
+#[derive(Debug, Serialize, PartialEq, Eq)]
+pub struct Content<'a> {
+    #[serde(rename = "type")]
     content_type: Option<&'a str>,
+
+    #[serde(rename = "value")]
+    value: Option<String>,
 }
 
-impl<'a> Sendgrid<'a> {
+#[derive(Debug, Serialize, PartialEq, Eq)]
+pub struct From {
+    #[serde(rename = "email")]
+    email: String,
+}
+
+#[derive(Debug, Serialize, PartialEq, Eq)]
+pub struct Personalization {
+    #[serde(rename = "to")]
+    to: Vec<From>,
+
+    #[serde(rename = "cc", skip_serializing_if = "Option::is_none")]
+    cc: Option<Vec<From>>,
+}
+
+impl<'a> SendgridEmail<'a> {
     #[must_use]
     /// Create a new sendgrid instance.
-    pub fn new(api_key: &'a str, to_email: &'a str, from_email: &'a str) -> Self {
+    pub fn new<T: AsRef<str>>(api_key: T, to_email: T, from_email: T) -> Self {
         Self {
-            api_key,
-            to_email,
-            from_email,
-            cc_emails: None,
-            content_type: None,
+            api_key: api_key.as_ref().to_owned(),
+            personalizations: vec![Personalization {
+                to: vec![From {
+                    email: to_email.as_ref().to_owned(),
+                }],
+                cc: None,
+            }],
+            from: From {
+                email: from_email.as_ref().to_owned(),
+            },
+            subject: None,
+            content: vec![Content {
+                content_type: None,
+                value: None,
+            }],
         }
     }
 
     /// Sends an email using Sendgrid API.
     /// # Example
     /// ```
-    /// use sendgrid_thin::Sendgrid;
-    /// let sendgrid = Sendgrid::new("SENDGRID_API_KEY", "to_email@example.com", "from_email@example.com");
+    /// use sendgrid_thin::SendgridEmail;
+    /// let mut sendgrid = SendgridEmail::new("SENDGRID_API_KEY", "to_email@example.com", "from_email@example.com");
     /// match sendgrid.send("subject", "body") {
     ///    Ok(_) => println!("Email sent successfully"),
     ///    Err(err) => println!("Error sending email: {}", err),
     /// }
     /// ```
-    pub fn send(&self, subject: &str, body: &str) -> Result<(), Box<dyn Error>> {
+    pub fn send<T: AsRef<str>>(&mut self, subject: T, body: T) -> Result<(), Box<dyn Error>> {
         ureq::post("https://api.sendgrid.com/v3/mail/send")
             .set("Authorization", &format!("Bearer {}", self.api_key))
             .set("Content-Type", "application/json")
-            .send_string(&self.set_body_email(subject, body))?;
+            .send_string(&self.set_body_email(subject.as_ref(), body.as_ref()))?;
```

Comment:
substituir refs por &

Reply:
This variable is a generic, it needs to be known as a `str`